### ISSUE TITLE
High/Increasing Operator CPU Usage

### ISSUE THREAD

**Issue by davidhiendl** [kubedb/project#798]
Hello,

I've noticed increasing CPU use when testing Kubedb community operator in at least the latest release version.
This is a fresh RKE cluster created solely for testing kubedb:
- Kubernetes Version: v1.19.4
- Provider: Hetzner Cloud, setup with RKE
- Other installed resources: Prometheus Operator + Instance
- hcloud/csi driver for persistent storage
- Kubedb + catalog installed via Helm into namespace: kube-addons (no other options set), Chart Version: v0.15.1

![Screenshot_20201123_162841](https://user-images.githubusercontent.com/11006964/99980834-fa2de700-2da8-11eb-9648-10a03cc5f3e8.png)

When inspecting the operator log it keeps trying to patch Elasticsearch/Mongodb statefulsets with updates to object orders.
This is a sample of operator logs:
```
I1123 15:19:17.909447       1 controller.go:113] Started processing, key: graylog/graylog-mongodb
I1123 15:19:17.911833       1 controller.go:113] Started processing, key: graylog/graylog-elasticsearch
I1123 15:19:18.009688       1 statefulset.go:81] Patching StatefulSet graylog/graylog-mongodb with {"spec":{"template":{"spec":{"securityContext":null}},"volumeClaimTemplates":[{"metadata":{"annotations":{"volume.beta.kubernetes.io/storage-class":"hcloud-volumes"},"creationTimestamp":null,"name":"datadir"},"spec":{"accessModes":["ReadWriteOnce"],"resources":{"requests":{"storage":"10Gi"}},"storageClassName":"hcloud-volumes"},"status":{}}]}}.
I1123 15:19:18.059726       1 service.go:77] Patching Service graylog/graylog-mongodb-stats with {"metadata":{"annotations":null}}.
I1123 15:19:18.137268       1 service.go:77] Patching Service graylog/graylog-mongodb-stats with {"metadata":{"annotations":{"monitoring.appscode.com/agent":"prometheus.io/operator"}}}.
I1123 15:19:18.396697       1 statefulset.go:81] Patching StatefulSet graylog/graylog-elasticsearch with {"spec":{"template":{"spec":{"$setElementOrder/containers":[{"name":"elasticsearch"},{"name":"exporter"}],"$setElementOrder/volumes":[{"name":"esconfig"},{"name":"temp-esconfig"},{"name":"transport-certs"},{"name":"temp"}],"containers":[{"$setElementOrder/env":[{"name":"ES_JAVA_OPTS"},{"name":"node.ingest"},{"name":"node.master"},{"name":"node.data"},{"name":"discovery.zen.minimum_master_nodes"},{"name":"node.name"},{"name":"cluster.name"},{"name":"network.host"},{"name":"ELASTIC_USER"},{"name":"ELASTIC_PASSWORD"},{"name":"discovery.zen.ping.unicast.hosts"}],"env":[{"name":"node.name","valueFrom":{"fieldRef":{"apiVersion":null}}}],"name":"elasticsearch"}],"volumes":[{"$retainKeys":["name","secret"],"name":"temp-esconfig","secret":{"defaultMode":null}},{"$retainKeys":["name","secret"],"name":"transport-certs","secret":{"defaultMode":null}}]}},"volumeClaimTemplates":[{"metadata":{"annotations":{"volume.beta.kubernetes.io/storage-class":"local-path"},"creationTimestamp":null,"name":"data"},"spec":{"accessModes":["ReadWriteOnce"],"resources":{"requests":{"storage":"50Gi"}},"storageClassName":"local-path"},"status":{}}]}}.
I1123 15:19:18.429859       1 service.go:77] Patching Service graylog/graylog-elasticsearch-stats with {"metadata":{"annotations":null}}.
I1123 15:19:18.491113       1 service.go:77] Patching Service graylog/graylog-elasticsearch-stats with {"metadata":{"annotations":{"monitoring.appscode.com/agent":"prometheus.io/operator"}}}.
I1123 15:19:18.649679       1 statefulset.go:81] Patching StatefulSet graylog/graylog-elasticsearch with {"spec":{"template":{"spec":{"$setElementOrder/containers":[{"name":"elasticsearch"},{"name":"exporter"}],"$setElementOrder/volumes":[{"name":"esconfig"},{"name":"temp-esconfig"},{"name":"transport-certs"},{"name":"temp"}],"containers":[{"$setElementOrder/env":[{"name":"ES_JAVA_OPTS"},{"name":"node.ingest"},{"name":"node.master"},{"name":"node.data"},{"name":"discovery.zen.minimum_master_nodes"},{"name":"node.name"},{"name":"cluster.name"},{"name":"network.host"},{"name":"ELASTIC_USER"},{"name":"ELASTIC_PASSWORD"},{"name":"discovery.zen.ping.unicast.hosts"}],"env":[{"name":"node.name","valueFrom":{"fieldRef":{"apiVersion":null}}}],"name":"elasticsearch"}],"volumes":[{"$retainKeys":["name","secret"],"name":"temp-esconfig","secret":{"defaultMode":null}},{"$retainKeys":["name","secret"],"name":"transport-certs","secret":{"defaultMode":null}}]}},"volumeClaimTemplates":[{"metadata":{"annotations":{"volume.beta.kubernetes.io/storage-class":"local-path"},"creationTimestamp":null,"name":"data"},"spec":{"accessModes":["ReadWriteOnce"],"resources":{"requests":{"storage":"50Gi"}},"storageClassName":"local-path"},"status":{}}]}}.
I1123 15:19:18.683686       1 service.go:77] Patching Service graylog/graylog-elasticsearch-stats with {"metadata":{"annotations":null}}.
I1123 15:19:18.741113       1 service.go:77] Patching Service graylog/graylog-elasticsearch-stats with {"metadata":{"annotations":{"monitoring.appscode.com/agent":"prometheus.io/operator"}}}.
I1123 15:19:22.096388       1 health.go:63] Starting health check for db graylog/graylog-mongodb
I1123 15:19:22.222932       1 health.go:71] Ending health check for db graylog/graylog-mongodb
I1123 15:19:22.223019       1 health.go:191] Ending health check loop
[xorm] [info]  2020/11/23 15:19:27.227313 PING DATABASE mysql
[xorm] [info]  2020/11/23 15:19:27.294397 PING DATABASE mysql
```

This is the deployed Elasticsearch CR:
```yaml
apiVersion: kubedb.com/v1alpha2
kind: Elasticsearch
metadata:
  name: graylog-elasticsearch
  namespace: graylog
spec:
  version: "6.8.10-xpack"
  replicas: 3
  storageType: Durable
  disableSecurity: false
  storage:
    storageClassName: local-path
    accessModes:
    - ReadWriteOnce
    resources:
      requests:
        storage: 50Gi
  enableSSL: false
  terminationPolicy: Halt
  monitor:
    agent: prometheus.io/operator
    prometheus:
      serviceMonitor:
        interval: 10s
        labels:
          monitor-group: k8s
  podTemplate:
    spec:
      resources:
        requests:
          memory: "1536Mi"
          cpu: "250m"
        limits:
          memory: "2048Mi"
          cpu: "750m"
```

I've also noticed that when patching the Elasticsearch/MySQL CRDs and changing the podTemplate memory requests it will not propagate to pods as it did with earlier versions of Kubedb.

**Comment by kamolhasan** [comment]
**CPU & Memory issue:**
 The new community operator is enriched with many new features and various bug fixes. To facilitate those features, new processes are running that weren't there before. So, you might notice a little jump in resource usage, but you can always control it by setting resource requests and limits.

**Change Propagation:**
 In the new community operator, the update strategy defaults to `OnDelete`. You need to delete the pods manually to see the changes. It provides users more flexibility like control over deletion order or performing prerequisite steps (ie. disable data allocation for safety etc.). The enterprise operator has a feature called "Smart Restart", to handle this.

**Comment by DrPhil** [comment]
We are seeing the same bad behaviour from the operator (in our case managing a singe MongoDB cluster). It is doing increasing amount of health checks towards the cluster until either:
a) the Mongo-pods fails due to the amount of open connections, or
b) the operator reaches its memory limit and is restarted

We are not seeing anything abnormal in the amount of logs from KubeDB, but can see increasing amount of opened connections from the operator in the mongo-pods logs, of these three types:
```
mongodb-0 mongodb 2020-12-17T11:23:26.068+0000 I NETWORK  [conn37840] received client metadata from 10.48.4.14:50874 conn37840: { driver: { name: "mongo-go-driver", version: "v1.3.2" }, os: { type: "linux", architecture: "amd64" }, platform: "go1.15.4" }
```
```
mongodb-2 mongodb 2020-12-17T11:23:26.067+0000 I NETWORK  [thread1] connection accepted from 10.48.4.14:40494 #38923 (3444 connections now open)
```
```
mongodb-2 mongodb 2020-12-17T11:23:26.125+0000 I -        [conn37445] end connection 10.48.4.14:54846 (3444 connections now open)
```

Here's some graphs showing the correlation between the number of logs from the Mongo-pods getting connections from the operator and the memory usage of the operator.

![image](https://user-images.githubusercontent.com/184889/102482685-e43bdb00-4063-11eb-8dcc-0bb5a01a963b.png)
![image](https://user-images.githubusercontent.com/184889/102482739-f9186e80-4063-11eb-80a8-957e152debb1.png)

**Comment by DrPhil** [comment]
We are seeing the same bad behaviour from the operator (in our case managing a singe MongoDB cluster). It is doing increasing amount of health checks towards the cluster until either:
a) the Mongo-pods fails due to the amount of open connections, or
b) the operator reaches its memory limit and is restarted

We are not seeing anything abnormal in the amount of logs from KubeDB, but can see increasing amount of opened connections from the operator in the mongo-pods logs, of these three types:
```
mongodb-0 mongodb 2020-12-17T11:23:26.068+0000 I NETWORK  [conn37840] received client metadata from 10.48.Y.XX:50874 conn37840: { driver: { name: "mongo-go-driver", version: "v1.3.2" }, os: { type: "linux", architecture: "amd64" }, platform: "go1.15.4" }
```
```
mongodb-2 mongodb 2020-12-17T11:23:26.067+0000 I NETWORK  [thread1] connection accepted from 10.48.Y.XX:40494 #38923 (3444 connections now open)
```
```
mongodb-2 mongodb 2020-12-17T11:23:26.125+0000 I -        [conn37445] end connection 10.48.Y.XX:54846 (3444 connections now open)
```

Here's some graphs showing the correlation between the number of logs from the Mongo-pods getting connections from the operator and the memory usage of the operator.

![image](https://user-images.githubusercontent.com/184889/102488358-35e86380-406c-11eb-8114-467f064dc1ae.png)
![Screenshot_2020-12-17_13-29-36](https://user-images.githubusercontent.com/184889/102488418-4c8eba80-406c-11eb-8dd1-b6f5fdb78d27.png)

**Comment by kamolhasan** [comment]
@faem can you take a look?

**Comment by davidhiendl** [comment]
The operator is most definitively leaking resources, but we have decided not to use KubeDB going forward.

**Comment by tamalsaha** [comment]
There was a memory leak is health checker. This issue has been resolved in KubeDB release v2021.03.17 . Please open a new bug if you still see this issue.
